serving/serving.go

```diff
@@ -82,13 +79,182 @@ func (serv *FeatureServer) TrainingData(req *pb.TrainingDataRequest, stream pb.F
+			}
+		default:
+			if err := serv.handleSplitDataRequest(&ctx); err != nil {
+				featureObserver.SetError()
+				return err
+			}
+		}
+	}
+}
+
+func (serv *FeatureServer) handleSplitInitializeRequest(ctx *splitContext) error {
+	ctx.logger.Infow("Initializing dataset", "id", ctx.req.Id, "shuffle", ctx.req.Shuffle, "testSize", ctx.req.TestSize)
+
+	trainTestSplitDef := provider.TrainTestSplitDef{
+		TrainingSetName:    ctx.req.Id.GetName(),
+		TrainingSetVariant: ctx.req.Id.GetVersion(),
+		TestSize:           ctx.req.TestSize,
+		Shuffle:            ctx.req.Shuffle,
+		RandomState:        int(ctx.req.RandomState),
+	}
+
+	cleanupFunc, err := serv.createTrainTestSplit(trainTestSplitDef)
+	defer cleanupFunc()
+	if err != nil {
+		return fferr.NewInternalError(err)
+	}
+	train, test, err := serv.getTrainTestSplitIterators(trainTestSplitDef)
+	if err != nil {
+		ctx.logger.Errorw("Failed to get training set iterator", "Error", err)
+		return err
+	}
+
+	*ctx.trainIterator = train
+	*ctx.testIterator = test
+
+	initResponse := &pb.BatchTrainTestSplitResponse{
+		RequestType: pb.RequestType_INITIALIZE,
+		Result:      &pb.BatchTrainTestSplitResponse_Initialized{Initialized: true},
+	}
+
+	if err := ctx.stream.Send(initResponse); err != nil {
+		ctx.logger.Errorw("Failed to send init response", "error", err)
+		return err
+	}
+
+	return nil
+}
+
+func (serv *FeatureServer) handleSplitDataRequest(ctx *splitContext) error {
+	var thisIter provider.TrainingSetIterator
+	switch ctx.req.GetRequestType() {
+	case pb.RequestType_TRAINING:
+		thisIter = *ctx.trainIterator
+	case pb.RequestType_TEST:
+		thisIter = *ctx.testIterator
+	default:
+		return fmt.Errorf("invalid request type")
+	}
+
+	rows := 0
+	trainingDataRows := make([]*pb.TrainingDataRow, 0)
+
+	for rows < int(ctx.req.BatchSize) {
+		if thisIter.Next() {
+			sRow, err := serializedRow(thisIter.Features(), thisIter.Label())
+			if err != nil {
+				return err
+			}
+			trainingDataRows = append(trainingDataRows, sRow)
+			rows++
+		} else {
+			// if we reach the end of the iterator mid-batch, we'll send the processed rows so far and end the iteration
+			serv.handleFinishedIterator(trainingDataRows, ctx)
+		}
+	}
+
+	response := &pb.BatchTrainTestSplitResponse{
+		Result: &pb.BatchTrainTestSplitResponse_Data{
+			Data: &pb.TrainingDataRows{Rows: trainingDataRows},
+		},
+	}
+
+	if err := ctx.stream.Send(response); err != nil {
+		ctx.logger.Errorw("Failed to write to stream", "Error", err)
+		return err
+	}
+
+	if thisIter.Err() != nil {
+		ctx.logger.Errorw("Dataset error", "Error", thisIter.Err())
+		return thisIter.Err()
+	}
+
+	return nil
+}
+
+func (serv *FeatureServer) handleFinishedIterator(trainingDataRows []*pb.TrainingDataRow, ctx *splitContext) {
+	if ctx.req.GetRequestType() == pb.RequestType_TEST {
+		*ctx.isTestFinished = true
+	} else if ctx.req.GetRequestType() == pb.RequestType_TRAINING {
+		*ctx.isTrainFinished = true
+	}
+
+	if *ctx.isTestFinished && *ctx.isTrainFinished {
```

Comment:
good q.

a request is either a testtype OR a traintype, never both, if one of them finishes we return an iterator finished response so that the iteration is stopped but the stream isn't closed